File: 二分查找/34.在排序数组中查找元素的第一个和最后一个位置.java
class Solution {

    public int[] searchRange(int[] nums, int target) {
        int left=find(nums,target);
        if(left== nums.length || nums[left]!=target) return new int[]{-1,-1};
        int right =find(nums,target+1)-1;
        return new int[]{left,right};
    }

    int find(int[] nums,int target){
        int left=-1,right=nums.length;
        while(left+1<right){
            int mid = (right+left)/2;
            if(nums[mid]<target){
                left=mid;
            }else{
                right=mid;
            }
        }
        return right;
    }
}

File: 二分查找/34.在排序数组中查找元素的第一个和最后一个位置.go
func searchRange(nums []int, target int) []int {
	pre := searchIndex(nums, target)
	if pre == len(nums) || nums[pre] != target {
		return []int{-1, -1}
	}
	last := searchIndex(nums, target+1)
	return []int{pre, last - 1}
}

func searchIndex(nums []int, target int) int {
	left, right := -1, len(nums)
	for left+1 < right {
		mid := left + (right-left)/2
		if nums[mid] < target {
			left = mid
		} else {
			right = mid
		}
	}
	return right
}